src/Development/IDE/Core/Shake.hs

```diff
@@ -268,30 +302,55 @@ instance NFData v => NFData (Value v)
 currentValue :: Value v -> Maybe v
 currentValue (Succeeded _ v) = Just v
 currentValue (Stale _ _) = Nothing
-currentValue Failed = Nothing
+currentValue Failed{} = Nothing
 
 -- | Return the most recent, potentially stale, value and a PositionMapping
 -- for the version of that value.
-lastValueIO :: ShakeExtras -> NormalizedFilePath -> Value v -> IO (Maybe (v, PositionMapping))
-lastValueIO ShakeExtras{positionMapping} file v = do
-    allMappings <- liftIO $ readVar positionMapping
-    pure $ case v of
-        Succeeded ver v -> Just (v, mappingForVersion allMappings file ver)
-        Stale ver v -> Just (v, mappingForVersion allMappings file ver)
-        Failed -> Nothing
+lastValueIO :: IdeRule k v => ShakeExtras -> k -> NormalizedFilePath -> IO (Maybe (v, PositionMapping))
+lastValueIO s@ShakeExtras{positionMapping,persistentKeys,state} k file = do
+    let readPersistent = do
+          pmap <- readVar persistentKeys
+          mv <- runMaybeT $ do
+            liftIO $ Logger.logInfo (logger s) $ T.pack $ "LOOKUP UP PERSISTENT FOR" ++ show k
+            f <- MaybeT $ pure $ HMap.lookup (Key k) pmap
+            dv <- MaybeT $ runIdeAction "lastValueIO" s $ f file
+            MaybeT $ pure $ fromDynamic dv
+          modifyVar state $ \hm -> case mv of
+            Nothing -> pure (HMap.insertWith upd (file,Key k) (Failed True) hm,Nothing)
+            Just v -> pure (HMap.insertWith upd (file,Key k) (Stale Nothing (toDyn v)) hm, Just (v,zeroMapping))
+
+        -- Update preserving 'monotonicity'
+        -- Don't want to overwrite a newer value with an older one
+        upd _new@(Failed False) old = old
+        upd new@(Failed True) (Failed False) = new
+        upd _new@(Failed True) old = old
+        upd new@(Stale _ _) Failed{} = new
+        upd new@(Stale v _) old@(Stale v' _) = if v >= v' then new else old
+        upd new@(Stale v _) old@(Succeeded v' _) = if v >= v' then new else old
+        upd new _old = new
```

Comment:
Could this be extracted to a `Monoid` instance?

Reply:
I think its best to keep this near the `insertWith` because any such monoid would be non-commutative, so it becomes confusing figuring out which the "new" one, and which is the "old" one.